### ISSUE TITLE
Improve error message for missing variable declaration

### ISSUE THREAD

**Issue by iamacup** [arangodb/arangodb#1805]
This query fails on my setup (details below) - please don't judge me on the horrible query - I was just messing about with how COLLECT statements worked when this broken. I don't need to know why or how to fix the query as i have a much better version working.

```
FOR v, e, p IN 2..2 OUTBOUND @personID with_job, at_company

	COLLECT company1 = v INTO jobs = p.vertices[1]

		FOR job IN jobs

			FOR paygrade, withpaygrade IN 1..1 OUTBOUND job._id with_paygrade
				
				COLLECT company2 = company1, paygrades = paygrade INTO data = {
					job: job,
					current: true,
					paygrades: {paygrade: paygrades, current: true}
				}

				RETURN {company: company2, current: true, jobs:data}
```

with this bind parameter (as per the web interface or php driver)

`{"personID": "people/90"}`

Error: `Internal error (location: arangod/Aql/AqlValue.cpp:182). Please report this error to arangodb.com (while executing) (location: arangod/RestHandler/RestCursorHandler.cpp:155). Please report this error to arangodb.com`

this does NOT fail:

```
FOR v, e, p IN 2..2 OUTBOUND @personID with_job, at_company

	COLLECT company1 = v INTO jobs = p.vertices[1]

		FOR job IN jobs

			FOR paygrade, withpaygrade IN 1..1 OUTBOUND job._id with_paygrade
				
				COLLECT company2 = company1 INTO data = {
					job: job,
					current: true,
					paygrades: {paygrade: paygrades, current: true}
				}

				RETURN {company: company2, current: true, jobs:data}
```

The difference is in the last COLLECT statement.

To make this super easy - i have attached the following:

1. arango-bare.zip - put the contents of this into a directory
2. arango-entry-zip - put the contents of this into a directory
3. docker-compose.yml

start the docker-compose file with `docker-compose up -d`, but edit it so that the two directories referenced by it are available (the two zip files)

EDIT THIS IN docker-compose.yml:

```
    volumes:
      - ./sliips/configuration-items/arango-bare:/opt/arango-bare
      - ./sliips/configuration-items/arango-entry:/opt/arango-entry
```

note - the arango-entry.zip contains the file that forces the container to run arangostore on the contents of the arango-bare directory once it is ready - basically once the container is up - go to the web console and run the query as above and it **should** produce the same error (heres hoping docker is as good as it claims)

[arango-entry.zip](https://github.com/arangodb/arangodb/files/222368/arango-entry.zip)
[arango-bare.zip](https://github.com/arangodb/arangodb/files/222370/arango-bare.zip) - this contains an arangodump of the whole DB & system collections
[docker-compose.yml.zip](https://github.com/arangodb/arangodb/files/222373/docker-compose.yml.zip)

<img width="1443" alt="error" src="https://cloud.githubusercontent.com/assets/12894620/14583277/220843c8-0415-11e6-8c14-d5fac8026ba6.png">

note - dw about the root password in the public domain.

**Comment by Simran-B** [comment]
In your first query, there are the variables `paygrade` and `withpaygrade`, the attribute keys `paygrades` and `paygrade` (I added quote marks for better distinction), but `paygrades` - used like a variable - is not declared anywhere (in light blue):

![1805](https://cloud.githubusercontent.com/assets/7819991/14584536/505fe5a6-044b-11e6-8c90-bdba0730f587.png)

Hence, the query is expected to fail. It should give a better understandable error message however, like
>use of undeclared identifier `paygrades`

**Comment by iamacup** [comment]
@Simran-B oops! but yeah - let's get a better, less ominous error message in there :)

**Comment by jsteemann** [comment]
There was an issue when a variable introduced in a COLLECT assignment (i.e. `paygrades` in the first query's `COLLECT company2 = company1, paygrades = paygrade`) was used by the COLLECT's INTO expression (i.e. ` INTO data = { job: job, current: true, paygrades: {paygrade: paygrades, current: true}`). 

In this case the behavior was undefined, which could make the query fail.
This is fixed in the 2.8 and devel branches. The fix will be shipped in 2.8.8.
